### ISSUE TITLE
New syntax for components which needs to be controlled

### ISSUE THREAD

**Issue by oyvindberg** [chandu0101/scalajs-react-components#26]
There is one pattern which really, really irritates me as a user. Take `MuiDialog`, for example.
Per now it requires the user to specify a `ref` string, look up that string through your components references, and if you're lucky and get scoping and strings right, you can call `open()` and `dismiss()` on it. Like this:

``` scala
val d   = MuiDialog(ref = "dialogref", ...)
val ref = Ref.toJS[MuiDialogM]("dialogref")

ref.foreach(_.show())
```

Try as I might, that fails to work a lot of the time. I discovered that the `ref` parameter to these components really has a type of `js.UndefinedOr[String | MuiDialogM ⇒ Unit]`, and suggest we expose that.

With that type signature we can write library/client code (not sure where it belongs - input wanted) like this:

``` scala
class DialogController{
  private var dialogRef: U[MuiDialogM] = uNone

  def setRef(d: MuiDialogM): Unit =
    dialogRef = d

  def invalidate(): Callback =
    Callback(dialogRef = uNone)

  val openDialog: Callback =
    Callback(dialogRef.foreach(_.show()))

  val closeDialog: Callback =
    Callback(dialogRef.foreach(_.dismiss()))
}
```

that you can use like this:

``` scala
MuiDialog(ref = dialogCtl.setRef _, ...)
...
MuiRaisedButton(onClick = (e: ReactEvent) ⇒ dialogCtl.openDialog)
```

So:
- if we do this, is there a need for the `String` version of the `ref` parameter at all?
- The code sample relies on a mutable `var`, can we do better?
- What kind of interface to things like this is best to expose to users?

**Comment by chandu0101** [comment]
> if we do this, is there a need for the String version of the ref parameter at all?

Short answer No, 
yeah react 0.13 suports call back refs  , once scalajs-react fully supports 0.13 and above we must remove all string refs from our code base, because in future version of react > 0.15/16/??? string refs becomes obsolete

here is the tweet from core react dev : 

> The next big (and probably most invasive change for a while) will be replacing string refs. Most can be automatically converted. #reactjs
> Use callback refs e.g.
> this.node = n} /> to avoid having to do any upgrade work. Allows you to move work from life-cycles too

.

> The code sample relies on a mutable var, can we do better?

currently i am also doing with mutable vars :( 

> What kind of interface to things like this is best to expose to users?

no ideas from me :( , hope we'll come up with a good solution  in future :)

**Comment by oyvindberg** [comment]
Oh, i hadnt realized it was a react 0.13 thing. in that case let's definitely wait until scalajs-react supports it properly. luckily, that seems scheduled for 0.10 now :)

**Comment by oyvindberg** [comment]
fixed as part of #40
